tests/validator/validator.cpp

```diff
@@ -2395,6 +2357,333 @@ TEST(Validator, unfoundUnitsInEncapsulatedComponents)
+
+    libcellml::ValidatorPtr validator = libcellml::Validator::create();
+
+    model->setName("model");
+    c1->setName("c1");
+    c2->setName("c2");
+    c3->setName("c3");
+
+    model->addComponent(c1);
+    model->addComponent(c2);
+    c2->addComponent(c3);
+
+    libcellml::VariablePtr v1 = libcellml::Variable::create();
+    v1->setName("v1");
+    v1->setUnits("dimensionless");
+
+    libcellml::VariablePtr v2 = libcellml::Variable::create();
+    v2->setName("v2");
+    v2->setUnits("dimensionless");
+
+    libcellml::VariablePtr v3 = libcellml::Variable::create();
+    v3->setName("v3");
+    v3->setUnits("dimensionless");
+
+    c1->addVariable(v1);
+    c2->addVariable(v2);
+    c3->addVariable(v3);
+
+    libcellml::Variable::addEquivalence(v1, v2);
+    libcellml::Variable::addEquivalence(v2, v3);
+
+    v1->setInterfaceType("public");
+    v2->setInterfaceType("public_and_private");
+    v3->setInterfaceType("public");
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->errorCount());
+}
+
+TEST(Validator, variableEquivalenceUnreachable)
+{
+    const std::vector<std::string> expectedErrors {
+        "The equivalence between 'v1' in component 'c1'  and 'v3' in component 'c3' is invalid. Component 'c1' and 'c3' are neither siblings nor in a parent/child relationship.",
+        "The equivalence between 'v3' in component 'c3'  and 'v1' in component 'c1' is invalid. Component 'c3' and 'c1' are neither siblings nor in a parent/child relationship.",
+    };
+
+    libcellml::ModelPtr model = libcellml::Model::create();
+    libcellml::ComponentPtr c1 = libcellml::Component::create();
+    libcellml::ComponentPtr c2 = libcellml::Component::create();
+    libcellml::ComponentPtr c3 = libcellml::Component::create();
+
+    libcellml::ValidatorPtr validator = libcellml::Validator::create();
+
+    model->setName("model");
+    c1->setName("c1");
+    c2->setName("c2");
+    c3->setName("c3");
+
+    model->addComponent(c1);
+    model->addComponent(c2);
+    c2->addComponent(c3);
+
+    libcellml::VariablePtr v1 = libcellml::Variable::create();
+    v1->setName("v1");
+    v1->setUnits("dimensionless");
+
+    libcellml::VariablePtr v2 = libcellml::Variable::create();
+    v2->setName("v2");
+    v2->setUnits("dimensionless");
+
+    libcellml::VariablePtr v3 = libcellml::Variable::create();
+    v3->setName("v3");
+    v3->setUnits("dimensionless");
+
+    c1->addVariable(v1);
+    c2->addVariable(v2);
+    c3->addVariable(v3);
+
+    // invalid equivalence, too far away
+    libcellml::Variable::addEquivalence(v1, v3);
+    validator->validateModel(model);
+
+    EXPECT_EQ_ERRORS(expectedErrors, validator);
+}
+
+TEST(Validator, variableEquivalenceUnreachableAndReachableTogether)
+{
+    const std::vector<std::string> expectedErrors = {
+        "Variable 'v1' has an interface type set to 'public' which is not the correct interface type for this variable. The interface type required is 'private'.",
```

Comment:
As an end-user, I have no idea what I could do with such an error message. I imagine that it's indirectly related to trying to create an equivalence between `v1` and `v2`, and then between `v2` and `v3`? Indeed, for those equivalences to work, I would expect `v1` to have a private interface.

Reply:
Seems pretty clear to me. `v1` has interface type `public` but it should be `private`.  I can only take away the the variable `v1` interface type should be set to `private`.

Reply:
Ok, sure, the message tells me what needs to be done to fix the problem, but that's the thing: I can't tell from the message what the problem is? Maybe some people don't care, but as an end-user I do or how can I avoid that problem in the future?

As I said earlier, and as far as I can tell, the problem is because of trying to create an equivalence between `v1` and `v2` and then between `v2` and `v3`. So, I would expect two error messages that reference those equivalences and then the fact that they are possible because of `v1` having a public interface rather than a private interface.

Reply:
This error has nothing to do with the equivalence between `v2` and `v3` and only to do with `v1` and `v2`.

I can only say that you would have to read the spec. for understanding.  I don't think it is desirable to say go and read the spec. my compiler doesn't tell me to go and read the standard every time I get something wrong.  I either get told that that is an error or that is an  error *did you mean ...*.

Reply:
> This error has nothing to do with the equivalence between `v2` and `v3` and only to do with `v1` and `v2`.

It indirectly has something to do with `v2` and `v3` since we need to go through `v1` and we can only do that if the interface is `private`.

> I can only say that you would have to read the spec. for understanding. I don't think it is desirable to say go and read the spec. my compiler doesn't tell me to go and read the standard every time I get something wrong. I either get told that that is an error or that is an error _did you mean ..._.

I did say what I would expect to see as an error, but we clearly seem to disagree. So be it, I will keep reviewing, but will leave it to others to approve. :)

Reply:
I think that (soon!) having a better place to point people towards will help here - that way they can read all about the problem and its solutions without having to have enormously long error messages returned.